File: Aman/DP Recursion and Backtracking/Dungeon Game.cpp
class Solution {
public:
    int calculateMinimumHP(vector<vector<int>>& dng) {
      int n=dng.size(),m=dng[0].size();
      int dp[n][m];
      for(int i=n-1;i>=0;i--)
      {
          for(int j=m-1;j>=0;j--)
          {
              if(i==n-1 && j==m-1)
                dp[i][j]=min(0,dng[i][j]);
              else if(i==n-1)
                dp[i][j]=min(0,dng[i][j]+dp[i][j+1]);
              else if(j==m-1)
                dp[i][j]=min(0,dng[i][j]+dp[i+1][j]);
              else
                dp[i][j]=min(0,max(dp[i][j+1],dp[i+1][j])+dng[i][j]);
          }
      }
      return abs(dp[0][0])+1;
    }
};

File: Harish Kumar/Dynamic Programming & Recursion/Dungeon Game/Dungeon Game.java
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        if(dungeon == null || dungeon.length == 0 || dungeon[0].length == 0)
        {
            return 0;
        }
        int dp[][] = new int[dungeon.length][dungeon[0].length];
        
        return helper(0,0, dungeon,dp);
        
    }
    private int helper(int sr, int sc, int dungeon[][], int dp[][])
    {
        if(sr >= dungeon.length || sc >= dungeon[0].length){
            return Integer.MAX_VALUE;
        }
        if(dp[sr][sc] != 0)
        {
            return dp[sr][sc];
        }
        int health = Math.min(helper(sr+1, sc, dungeon,dp), helper(sr, sc+1, dungeon,dp));
        
        if(health == Integer.MAX_VALUE)
        {
            health = 1;
        }
        
        int res =  0;          //Math.max(health - dungeon[sr][sc],1);
        if(health - dungeon[sr][sc] > 0)
        {
            res = health - dungeon[sr][sc];
            dp[sr][sc] = res;
        }
        else
        {
            res = 1;
            dp[sr][sc] = 1;
        }
            
        return res;
    }
}


